File: codes/go/claude/complete/task_4/Solution.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
)

func allocateResources(patients [][]int, resources []map[string]interface{}, responders int, totalTime int) map[string]interface{} {
	assignments := []map[string]interface{}{}
	survivors := 0
	totalSurvivalProbability := 0.0

	dp := make([][][]float64, totalTime+1)
	for t := range dp {
		dp[t] = make([][]float64, len(patients))
		for p := range dp[t] {
			dp[t][p] = make([]float64, responders+1)
		}
	}

	for t := 1; t <= totalTime; t++ {
		for p := 0; p < len(patients); p++ {
			for r := 1; r <= responders; r++ {
				dp[t][p][r] = dp[t-1][p][r]
				if t < len(patients[p]) {
					dp[t][p][r] = max(dp[t][p][r], float64(patients[p][t]))
				}
				for _, resource := range resources {
					resourceTime := int(resource["time"].(float64))
					resourceBoost := resource["boost"].(float64)
					if t >= resourceTime {
						prevProb := dp[t-resourceTime][p][r-1]
						if prevProb > 0 {
							boostedProb := min(prevProb+resourceBoost, 100)
							if boostedProb > dp[t][p][r] {
								dp[t][p][r] = boostedProb
							}
						}
					}
				}
			}
		}
	}

	assigned := make([]bool, len(patients))
	for t := totalTime; t > 0; t-- {
		for r := responders; r > 0; r-- {
			for p := 0; p < len(patients); p++ {
				if !assigned[p] && dp[t][p][r] > dp[t][p][r-1] {
					for _, resource := range resources {
						resourceTime := int(resource["time"].(float64))
						resourceBoost := resource["boost"].(float64)
						if t >= resourceTime {
							prevProb := dp[t-resourceTime][p][r-1]
							boostedProb := min(prevProb+resourceBoost, 100)
							if boostedProb == dp[t][p][r] {
								assignments = append(assignments, map[string]interface{}{
									"patient":  p,
									"resource": resource,
									"time":     t - resourceTime,
								})
								assigned[p] = true
								break
							}
						}
					}
					break
				}
			}
		}
	}

	for p, prob := range dp[totalTime] {
		finalProb := prob[responders]
		if finalProb > 50 {
			survivors++
		}
		totalSurvivalProbability += finalProb
	}

	return map[string]interface{}{
		"assignments":               assignments,
		"survivors":                 survivors,
		"total_survival_probability": totalSurvivalProbability,
	}
}

func max(a, b float64) float64 {
	if a > b {
		return a
	}
	return b
}

func min(a, b float64) float64 {
	if a < b {
		return a
	}
	return b
}

func main() {
	var input map[string]interface{}
	decoder := json.NewDecoder(os.Stdin)
	err := decoder.Decode(&input)
	if err != nil {
		fmt.Println("Error decoding input:", err)
		return
	}

	patients := input["patients"].([]interface{})
	patientsProbabilities := make([][]int, len(patients))
	for i, patient := range patients {
		probabilities := patient.([]interface{})
		patientsProbabilities[i] = make([]int, len(probabilities))
		for j, prob := range probabilities {
			patientsProbabilities[i][j] = int(prob.(float64))
		}
	}

	resources := input["resources"].([]interface{})
	resourcesMaps := make([]map[string]interface{}, len(resources))
	for i, resource := range resources {
		resourcesMaps[i] = resource.(map[string]interface{})
	}

	responders := int(input["responders"].(float64))
	totalTime := int(input["total_time"].(float64))

	result := allocateResources(patientsProbabilities, resourcesMaps, responders, totalTime)

	encoder := json.NewEncoder(os.Stdout)
	encoder.SetIndent("", "  ")
	err = encoder.Encode(result)
	if err != nil {
		fmt.Println("Error encoding result:", err)
	}
}

File: codes/javascript/claude/complete/task_4/Solution.js
const readline = require('readline');
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
  terminal: false
});

let inputData = '';

rl.on('line', (line) => {
  inputData += line;
});

rl.on('close', () => {
  const input = JSON.parse(inputData);
  const result = allocateResources(input.patients, input.resources, input.responders, input.total_time);
  console.log(JSON.stringify(result));
});

function allocateResources(patients, resources, responders, totalTime) {
  const assignments = [];
  let survivors = 0;
  let totalSurvivalProbability = 0;

  const dp = new Array(totalTime + 1).fill(null).map(() => 
    new Array(responders + 1).fill(null).map(() => 
      new Array(patients.length).fill(0)
    )
  );

  const backtrack = new Array(totalTime + 1).fill(null).map(() => 
    new Array(responders + 1).fill(null).map(() => 
      new Array(patients.length).fill(null)
    )
  );

  for (let t = 0; t <= totalTime; t++) {
    for (let r = 0; r <= responders; r++) {
      for (let p = 0; p < patients.length; p++) {
        if (t === 0 || r === 0) {
          dp[t][r][p] = p === 0 ? patients[p][0] : dp[t][r][p-1];
          continue;
        }

        dp[t][r][p] = Math.max(
          p > 0 ? dp[t][r][p-1] : 0,
          dp[t][r-1][p]
        );

        for (let i = 0; i < resources.length; i++) {
          const resource = resources[i];
          if (t >= resource.time && p > 0) {
            const prevProb = dp[t-resource.time][r-1][p-1];
            const patientProb = patients[p][t-resource.time];
            const newProb = Math.min(100, patientProb + resource.boost);
            const totalProb = prevProb + newProb;

            if (totalProb > dp[t][r][p]) {
              dp[t][r][p] = totalProb;
              backtrack[t][r][p] = {patient: p, resource: i, time: t - resource.time};
            }
          }
        }
      }
    }
  }

  let t = totalTime;
  let r = responders;
  let p = patients.length - 1;

  while (t > 0 && r > 0 && p >= 0) {
    const assignment = backtrack[t][r][p];
    if (assignment) {
      assignments.unshift(assignment);
      t = assignment.time;
      r--;
      p--;
    } else {
      p--;
    }
  }

  const finalProbabilities = new Array(patients.length).fill(0);

  for (const assignment of assignments) {
    const { patient, resource, time } = assignment;
    const baseProb = patients[patient][time];
    finalProbabilities[patient] = Math.min(100, baseProb + resources[resource].boost);
  }

  for (let i = 0; i < patients.length; i++) {
    if (finalProbabilities[i] === 0) {
      finalProbabilities[i] = patients[i][totalTime];
    }
    if (finalProbabilities[i] > 50) {
      survivors++;
    }
    totalSurvivalProbability += finalProbabilities[i];
  }

  return {
    assignments,
    survivors,
    total_survival_probability: totalSurvivalProbability
  };
}